src/debug_sdfs_ed64.c

```diff
@@ -0,0 +1,590 @@
+#define ED64_SD_CMD12                   0x4C // Stop transmission on multiple block read
+#define ED64_SD_CMD18                   0x52 // Read multiple block
+#define ED64_SD_CMD25                   0x59 // Write multiple block
+#define ED64_SD_CMD55                   0x77 // Application specific cmd
+#define ED64_SD_CMD41                   0x69
+
+// Everdrive SD implementation state bits
+#define ED64_SD_MODE_ACCESS                 0x03
+#define ED64_SD_MODE_COMM                   0x0C
+// We only support verion > 2.0, so this is just on/off
+#define ED64_SD_MODE_IS_HC                  0x40
+
+// Everdrive SD data access mode
+#define ED64_SD_MODE_NONE                   0x00
+#define ED64_SD_MODE_BLOCK_READ             0x01
+#define ED64_SD_MODE_BLOCK_WRITE            0x02
+
+// Everdrive communication mode - see everdrive_sd_set_mode
+#define ED64_SD_MODE_CMD_READ           0x00
+#define ED64_SD_MODE_CMD_WRITE          0x04
+#define ED64_SD_MODE_DATA_READ          0x08
+#define ED64_SD_MODE_DATA_WRITE         0x0C
+
+static uint32_t everdrive_sd_active_mode = 0;
+
+// Set the mode to talk to the SD card. In ED64_SD_MODE_CMD_X modes, the
+// bytes are read from/written to the cmd line. WithED64_SD_MODE_DATA_X,
+// the provided bytes are written in SD wide bus format. e.g for [abcd efgh],
+// the output on the 4 data lanes will look like;
+// dat3: ae
+// dat2: bf
+// dat1: cg
+// dat0: dh
+// Effectively, the clock line will be toggled by the amount set by
+// set_everdrive_sd_bitlen for each read/write. For example on the data mode,
+// setting the bit len to 2 will output a single byte for every
+// everdrive_sd_write
+static void everdrive_sd_set_mode(uint8_t mode) {
+	if ((everdrive_sd_active_mode & ED64_SD_MODE_COMM) == mode) return;
+	everdrive_sd_active_mode &= ~ED64_SD_MODE_COMM;
+	everdrive_sd_active_mode |= mode;
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + mode, 0xffff);
+}
+
+static uint32_t __attribute__((aligned(16))) everdrive_sd_config = 0;
+// Sets how many bits are read/written at a time, per lane. This is equal to the
+// times the clock line is toggled per read/write. The actual data manipulated
+// depends on the mode (see everdrive_sd_set_mode)
+static void set_everdrive_sd_bitlen(uint8_t val) {
+	everdrive_sd_config &= ~ED64_SD_CFG_BITLEN;
+	everdrive_sd_config |= (val & ED64_SD_CFG_BITLEN);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_STATUS, everdrive_sd_config);
+}
+
+void everdrive_sd_busy() {
+	while ((io_read(ED64_BASE_ADDRESS + ED64_REGISTER_SD_STATUS)
+		& ED64_SD_STATUS_BUSY) != 0);
+}
+
+void everdrive_sd_write_command(uint8_t val) {
+	everdrive_sd_set_mode(ED64_SD_MODE_CMD_WRITE);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_CMD_WRITE, val);
+	everdrive_sd_busy();
+}
+
+uint8_t everdrive_sd_read_command() {
+	everdrive_sd_set_mode(ED64_SD_MODE_CMD_READ);
+	io_write(ED64_BASE_ADDRESS + ED64_REGISTER_SD_BASE + ED64_SD_MODE_CMD_READ, 0xffff);
+	everdrive_sd_busy();
```

Comment:
It's not clear to me why here you need to check for the busy flag. When is this busy check required?

Reply:
I honestly do not know. This is part of the hardware design and I was not able to decipher this. Seems like it is only required for cmd read/write